File: supabase/functions/_shared/regulation_document.ts
import type { RegulationField } from './regulation.ts';

export const REGULATION_DOCUMENT_VERSION = 1;

export const REGULATION_SECTION_CODES = [
  'eligibility',
  'schedule_venue',
  'registration_payment',
  'match_operations',
  'awards',
  'refund_changes',
  'notices_contact',
  'other',
] as const;

export type RegulationSectionCode = typeof REGULATION_SECTION_CODES[number];
export type RegulationAvailability = 'present' | 'not_announced' | 'not_applicable';
export type RegulationBlockType =
  | 'paragraph'
  | 'subheading'
  | 'bullets'
  | 'key_values'
  | 'table'
  | 'notice'
  | 'division_schedule';

export interface RegulationEntry {
  label: string;
  value: string;
}

export interface RegulationTableRow {
  cells: string[];
}

export interface RegulationDivisionItem {
  name: string;
  date?: string;
  venue?: string;
  fee?: string;
  account?: string;
  capacity?: string;
}

export interface RegulationBlock {
  type: RegulationBlockType;
  text?: string;
  items?: string[];
  entries?: RegulationEntry[];
  columns?: string[];
  rows?: RegulationTableRow[];
  divisions?: RegulationDivisionItem[];
}

export interface RegulationSection {
  code: RegulationSectionCode;
  availability: RegulationAvailability;
  blocks: RegulationBlock[];
}

export interface RegulationDocument {
  schema_version: typeof REGULATION_DOCUMENT_VERSION;
  summary?: string;
  sections: RegulationSection[];
}

const SECTION_LABELS: Record<RegulationSectionCode, string> = {
  eligibility: '참가 부서 및 자격',
  schedule_venue: '일정 및 장소',
  registration_payment: '신청 및 결제',
  match_operations: '경기 방식 및 운영',
  awards: '시상 및 참가상품',
  refund_changes: '변경·취소·환불',
  notices_contact: '유의사항 및 문의',
  other: '기타 안내',
};

const BLOCK_TYPES = new Set<RegulationBlockType>([
  'paragraph',
  'subheading',
  'bullets',
  'key_values',
  'table',
  'notice',
  'division_schedule',
]);
const AVAILABILITIES = new Set<RegulationAvailability>([
  'present',
  'not_announced',
  'not_applicable',
]);
const SECTION_CODE_SET = new Set<RegulationSectionCode>(REGULATION_SECTION_CODES);

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function cleanText(value: unknown, max: number): string | undefined {
  if (typeof value !== 'string') return undefined;
  const text = value.replace(/\s+/g, ' ').trim();
  if (!text) return undefined;
  return text.length <= max ? text : text.slice(0, max).trimEnd();
}

function cleanLines(value: unknown, limit: number, max: number): string[] {
  if (!Array.isArray(value)) return [];
  return value
    .map((item) => cleanText(item, max))
    .filter((item): item is string => item !== undefined)
    .slice(0, limit);
}

function normalizeEntries(value: unknown): RegulationEntry[] {
  if (!Array.isArray(value)) return [];
  const out: RegulationEntry[] = [];
  for (const item of value.slice(0, 30)) {
    if (!isRecord(item)) continue;
    const label = cleanText(item.label, 80);
    const entryValue = cleanText(item.value, 1000);
    if (label && entryValue) out.push({ label, value: entryValue });
  }
  return out;
}

function normalizeRows(value: unknown): RegulationTableRow[] {
  if (!Array.isArray(value)) return [];
  const out: RegulationTableRow[] = [];
  for (const item of value.slice(0, 50)) {
    if (!isRecord(item)) continue;
    const cells = cleanLines(item.cells, 8, 500);
    if (cells.length > 0) out.push({ cells });
  }
  return out;
}

function normalizeDivisions(value: unknown): RegulationDivisionItem[] {
  if (!Array.isArray(value)) return [];
  const out: RegulationDivisionItem[] = [];
  for (const item of value.slice(0, 30)) {
    if (!isRecord(item)) continue;
    const name = cleanText(item.name, 120);
    if (!name) continue;
    out.push({
      name,
      date: cleanText(item.date, 160),
      venue: cleanText(item.venue, 300),
      fee: cleanText(item.fee, 200),
      account: cleanText(item.account, 300),
      capacity: cleanText(item.capacity, 160),
    });
  }
  return out;
}

function normalizeBlock(value: unknown): RegulationBlock | null {
  if (!isRecord(value) || typeof value.type !== 'string') return null;
  if (!BLOCK_TYPES.has(value.type as RegulationBlockType)) return null;
  const type = value.type as RegulationBlockType;

  switch (type) {
    case 'paragraph':
    case 'notice': {
      const text = cleanText(value.text, 2000);
      return text ? { type, text } : null;
    }
    case 'subheading': {
      const text = cleanText(value.text, 160);
      return text ? { type, text } : null;
    }
    case 'bullets': {
      const items = cleanLines(value.items, 50, 800);
      return items.length > 0 ? { type, items } : null;
    }
    case 'key_values': {
      const entries = normalizeEntries(value.entries);
      return entries.length > 0 ? { type, entries } : null;
    }
    case 'table': {
      const columns = cleanLines(value.columns, 8, 120);
      const rows = normalizeRows(value.rows);
      return rows.length > 0 ? { type, columns, rows } : null;
    }
    case 'division_schedule': {
      const divisions = normalizeDivisions(value.divisions);
      return divisions.length > 0 ? { type, divisions } : null;
    }
  }
}

export function normalizeRegulationDocument(raw: unknown): RegulationDocument | null {
  if (!isRecord(raw) || raw.schema_version !== REGULATION_DOCUMENT_VERSION) return null;
  if (!Array.isArray(raw.sections)) return null;

  const byCode = new Map<RegulationSectionCode, RegulationSection>();
  for (const item of raw.sections) {
    if (!isRecord(item) || typeof item.code !== 'string') continue;
    if (!SECTION_CODE_SET.has(item.code as RegulationSectionCode)) continue;
    const code = item.code as RegulationSectionCode;
    const availability = typeof item.availability === 'string' &&
        AVAILABILITIES.has(item.availability as RegulationAvailability)
      ? item.availability as RegulationAvailability
      : 'present';
    const blocks = Array.isArray(item.blocks)
      ? item.blocks.map(normalizeBlock).filter((block): block is RegulationBlock => block !== null)
        .slice(0, 20)
      : [];
    if (availability === 'present' && blocks.length === 0) continue;

    const existing = byCode.get(code);
    if (existing && existing.availability === 'present' && availability === 'present') {
      existing.blocks.push(...blocks);
      existing.blocks = existing.blocks.slice(0, 20);
    } else if (!existing || availability === 'present') {
      byCode.set(code, { code, availability, blocks });
    }
  }

  const sections = REGULATION_SECTION_CODES
    .map((code) => byCode.get(code))
    .filter((section): section is RegulationSection => section !== undefined);
  if (sections.length === 0) return null;
  return {
    schema_version: REGULATION_DOCUMENT_VERSION,
    summary: cleanText(raw.summary, 500),
    sections,
  };
}

export function regulationSectionLabel(code: RegulationSectionCode): string {
  return SECTION_LABELS[code];
}

function normalizedLegacyLabel(label: string): string {
  return label.toLowerCase().replace(/[\s_\-·:/()]/g, '');
}

export function sectionCodeForLegacyLabel(label: string): RegulationSectionCode {
  const value = normalizedLegacyLabel(label);
  if (/참가부서|참가자격|출전규정|예외부서|시드기준|참가규모|경기종목/.test(value)) {
    return 'eligibility';
  }
  if (/일시|일정|대회일|경기일|장소|경기장|개최지/.test(value)) {
    return 'schedule_venue';
  }
  if (/환불|취소|변경/.test(value)) return 'refund_changes';
  if (/신청|접수|참가비|입금|계좌|결제|예금주/.test(value)) {
    return 'registration_payment';
  }
  if (/시상|상금|참가상품|기념품/.test(value)) return 'awards';
  if (/경기방식|진행방식|운영|사용구|주최|주관|후원|협찬/.test(value)) {
    return 'match_operations';
  }
  if (/문의|연락|전화|담당|안내/.test(value)) return 'notices_contact';
  return 'other';
}

export function regulationDocumentFromLegacy(
  fields: RegulationField[],
  notes: string[] = [],
): RegulationDocument | null {
  const grouped = new Map<RegulationSectionCode, RegulationEntry[]>();
  for (const field of fields) {
    const label = cleanText(field.label, 80);
    const value = cleanText(field.value, 1000);
    if (!label || !value) continue;
    const code = sectionCodeForLegacyLabel(label);
    const entries = grouped.get(code) ?? [];
    entries.push({ label, value });
    grouped.set(code, entries);
  }

  const sections: RegulationSection[] = [];
  for (const code of REGULATION_SECTION_CODES) {
    const entries = grouped.get(code);
    if (entries?.length) {
      sections.push({ code, availability: 'present', blocks: [{ type: 'key_values', entries }] });
    }
  }
  const cleanNotes = notes.map((note) => cleanText(note, 800)).filter((note): note is string =>
    note !== undefined
  );
  if (cleanNotes.length > 0) {
    const existing = sections.find((section) => section.code === 'notices_contact');
    const noticeBlock: RegulationBlock = { type: 'bullets', items: cleanNotes.slice(0, 50) };
    if (existing) existing.blocks.push(noticeBlock);
    else {
      sections.push({ code: 'notices_contact', availability: 'present', blocks: [noticeBlock] });
      sections.sort((a, b) =>
        REGULATION_SECTION_CODES.indexOf(a.code) - REGULATION_SECTION_CODES.indexOf(b.code)
      );
    }
  }
  return sections.length > 0 ? { schema_version: REGULATION_DOCUMENT_VERSION, sections } : null;
}

export function mergeRegulationDocuments(
  preferred: RegulationDocument | null,
  supplement: RegulationDocument,
): RegulationDocument {
  if (!preferred) return supplement;
  const preferredByCode = new Map(preferred.sections.map((section) => [section.code, section]));
  const supplementByCode = new Map(supplement.sections.map((section) => [section.code, section]));
  const sections = REGULATION_SECTION_CODES.map((code) =>
    preferredByCode.get(code) ?? supplementByCode.get(code)
  ).filter((section): section is RegulationSection => section !== undefined);
  return {
    schema_version: REGULATION_DOCUMENT_VERSION,
    summary: supplement.summary ?? preferred.summary,
    sections,
  };
}

export function regulationDocumentToFields(document: RegulationDocument): RegulationField[] {
  const out: RegulationField[] = [];
  for (const section of document.sections) {
    if (section.availability !== 'present') continue;
    for (const block of section.blocks) {
      if (block.type === 'key_values') out.push(...(block.entries ?? []));
      if (block.type === 'division_schedule') {
        const value = (block.divisions ?? []).map((division) =>
          [
            division.name,
            division.date,
            division.venue,
            division.fee,
            division.account,
            division.capacity,
          ]
            .filter((part): part is string => Boolean(part))
            .join(' · ')
        ).join('\n');
        if (value) out.push({ label: '부서별 일정·장소', value });
      }
    }
  }
  return out;
}

export function regulationDocumentToNotes(document: RegulationDocument): string[] {
  const section = document.sections.find((item) => item.code === 'notices_contact');
  if (!section || section.availability !== 'present') return [];
  const out: string[] = [];
  for (const block of section.blocks) {
    if (block.type === 'notice' && block.text) out.push(block.text);
    if (block.type === 'bullets') out.push(...(block.items ?? []));
  }
  return out;
}

function blockLines(block: RegulationBlock): string[] {
  switch (block.type) {
    case 'subheading':
      return block.text ? [`◈ ${block.text}`] : [];
    case 'paragraph':
      return block.text ? [block.text] : [];
    case 'notice':
      return block.text ? [`※ ${block.text}`] : [];
    case 'bullets':
      return (block.items ?? []).map((item) => `- ${item}`);
    case 'table':
      return [
        ...((block.columns ?? []).length > 0 ? [(block.columns ?? []).join(' | ')] : []),
        ...(block.rows ?? []).map((row) => row.cells.join(' | ')),
      ];
    case 'key_values':
    case 'division_schedule':
      return [];
  }
}

export function regulationDocumentToBody(document: RegulationDocument): string {
  const lines: string[] = [];
  for (const section of document.sections) {
    if (section.availability !== 'present') continue;
    const content = section.blocks.flatMap(blockLines);
    if (content.length === 0) continue;
    lines.push(`● ${SECTION_LABELS[section.code]}`, ...content);
  }
  return lines.join('\n').trim();
}

export function regulationDocumentVerificationFields(
  document: RegulationDocument,
): RegulationField[] {
  const out: RegulationField[] = [];
  for (const section of document.sections) {
    if (section.availability !== 'present') continue;
    // 문의처 label만 전화번호 오탐 예외를 적용한다. notices_contact 전체를
    // "문의"로 넘기면 그 섹션의 금액·날짜까지 원문 검증을 우회할 수 있다.
    const sectionLabel = section.code === 'notices_contact'
      ? '유의사항'
      : SECTION_LABELS[section.code];
    for (const block of section.blocks) {
      if (block.text) out.push({ label: sectionLabel, value: block.text });
      for (const item of block.items ?? []) out.push({ label: sectionLabel, value: item });
      out.push(...(block.entries ?? []));
      for (const row of block.rows ?? []) {
        out.push({ label: sectionLabel, value: row.cells.join(' ') });
      }
      for (const division of block.divisions ?? []) {
        out.push({
          label: sectionLabel,
          value: [
            division.name,
            division.date,
            division.venue,
            division.fee,
            division.account,
            division.capacity,
          ].filter((part): part is string => Boolean(part)).join(' '),
        });
      }
    }
  }
  return out;
}

export function regulationDocumentText(document: RegulationDocument): string {
  const lines: string[] = [];
  if (document.summary) lines.push(document.summary);
  for (const section of document.sections) {
    if (section.availability !== 'present') continue;
    lines.push(SECTION_LABELS[section.code]);
    for (const block of section.blocks) {
      if (block.text) lines.push(block.text);
      lines.push(...(block.items ?? []));
      for (const entry of block.entries ?? []) lines.push(`${entry.label}: ${entry.value}`);
      if ((block.columns ?? []).length > 0) lines.push((block.columns ?? []).join(' | '));
      for (const row of block.rows ?? []) lines.push(row.cells.join(' | '));
      for (const division of block.divisions ?? []) {
        lines.push(
          [
            division.name,
            division.date,
            division.venue,
            division.fee,
            division.account,
            division.capacity,
          ].filter((part): part is string => Boolean(part)).join(' · '),
        );
      }
    }
  }
  return lines.join('\n');
}

File: app/lib/models/regulation_document.dart
import 'regulation_body_lines.dart';

enum RegulationSectionCode {
  eligibility('eligibility', '참가 부서 및 자격'),
  scheduleVenue('schedule_venue', '일정 및 장소'),
  registrationPayment('registration_payment', '신청 및 결제'),
  matchOperations('match_operations', '경기 방식 및 운영'),
  awards('awards', '시상 및 참가상품'),
  refundChanges('refund_changes', '변경·취소·환불'),
  noticesContact('notices_contact', '유의사항 및 문의'),
  other('other', '기타 안내');

  const RegulationSectionCode(this.value, this.label);

  final String value;
  final String label;

  static RegulationSectionCode? tryParse(Object? value) {
    if (value is! String) return null;
    for (final code in values) {
      if (code.value == value) return code;
    }
    return null;
  }
}

enum RegulationAvailability {
  present,
  notAnnounced,
  notApplicable;

  static RegulationAvailability? tryParse(Object? value) => switch (value) {
        'present' => RegulationAvailability.present,
        'not_announced' => RegulationAvailability.notAnnounced,
        'not_applicable' => RegulationAvailability.notApplicable,
        _ => null,
      };
}

enum RegulationBlockType {
  paragraph,
  subheading,
  bullets,
  keyValues,
  table,
  notice,
  divisionSchedule;

  static RegulationBlockType? tryParse(Object? value) => switch (value) {
        'paragraph' => RegulationBlockType.paragraph,
        'subheading' => RegulationBlockType.subheading,
        'bullets' => RegulationBlockType.bullets,
        'key_values' => RegulationBlockType.keyValues,
        'table' => RegulationBlockType.table,
        'notice' => RegulationBlockType.notice,
        'division_schedule' => RegulationBlockType.divisionSchedule,
        _ => null,
      };
}

class RegulationDocument {
  const RegulationDocument({
    required this.schemaVersion,
    required this.sections,
    this.summary,
  });

  static const currentSchemaVersion = 1;

  final int schemaVersion;
  final String? summary;
  final List<RegulationSection> sections;

  bool get isEmpty => sections.isEmpty;

  static RegulationDocument? tryFromJson(Object? raw) {
    final json = _map(raw);
    if (json == null || json['schema_version'] != currentSchemaVersion) {
      return null;
    }
    final rawSections = json['sections'];
    if (rawSections is! List) return null;

    final byCode = <RegulationSectionCode, RegulationSection>{};
    for (final rawSection in rawSections) {
      final section = RegulationSection.tryFromJson(rawSection);
      if (section == null) continue;
      byCode.putIfAbsent(section.code, () => section);
    }
    final sections = RegulationSectionCode.values
        .map((code) => byCode[code])
        .whereType<RegulationSection>()
        .toList(growable: false);
    if (sections.isEmpty) return null;
    return RegulationDocument(
      schemaVersion: currentSchemaVersion,
      summary: _text(json['summary']),
      sections: sections,
    );
  }

  static RegulationDocument? fromLegacy({
    required List<({String label, String value})> fields,
    required List<String> notes,
    String? body,
  }) {
    final grouped = <RegulationSectionCode, List<RegulationEntry>>{};
    for (final field in fields) {
      final label = field.label.trim();
      final value = field.value.trim();
      if (label.isEmpty || value.isEmpty) continue;
      final code = sectionCodeForLegacyLabel(label);
      grouped.putIfAbsent(code, () => []).add(
            RegulationEntry(label: label, value: value),
          );
    }

    final sections = <RegulationSection>[];
    for (final code in RegulationSectionCode.values) {
      final entries = grouped[code];
      if (entries == null || entries.isEmpty) continue;
      sections.add(
        RegulationSection(
          code: code,
          availability: RegulationAvailability.present,
          blocks: [
            RegulationBlock(
              type: RegulationBlockType.keyValues,
              entries: entries,
            ),
          ],
        ),
      );
    }

    final cleanNotes = notes
        .map((note) => note.trim())
        .where((note) => note.isNotEmpty)
        .toList();
    if (cleanNotes.isNotEmpty) {
      _appendBlock(
        sections,
        RegulationSectionCode.noticesContact,
        RegulationBlock(type: RegulationBlockType.bullets, items: cleanNotes),
      );
    }
    final cleanBody = body?.trim();
    if (cleanBody != null && cleanBody.isNotEmpty) {
      for (final block in _legacyBodyBlocks(cleanBody)) {
        _appendBlock(sections, RegulationSectionCode.other, block);
      }
    }
    sections.sort((a, b) => a.code.index.compareTo(b.code.index));
    if (sections.isEmpty) return null;
    return RegulationDocument(
      schemaVersion: currentSchemaVersion,
      sections: List.unmodifiable(sections),
    );
  }

  static void _appendBlock(
    List<RegulationSection> sections,
    RegulationSectionCode code,
    RegulationBlock block,
  ) {
    final index = sections.indexWhere((section) => section.code == code);
    if (index < 0) {
      sections.add(
        RegulationSection(
          code: code,
          availability: RegulationAvailability.present,
          blocks: [block],
        ),
      );
      return;
    }
    final section = sections[index];
    sections[index] = RegulationSection(
      code: section.code,
      availability: section.availability,
      blocks: [...section.blocks, block],
    );
  }
}

List<RegulationBlock> _legacyBodyBlocks(String body) {
  final blocks = <RegulationBlock>[];
  for (final line in publicRegulationBodyLines(body)) {
    switch (line.kind) {
      case RegulationLineKind.header:
      case RegulationLineKind.item:
        blocks.add(
          RegulationBlock(
            type: RegulationBlockType.subheading,
            text: line.text,
          ),
        );
        break;
      case RegulationLineKind.numbered:
        blocks.add(
          RegulationBlock(
            type: RegulationBlockType.bullets,
            items: ['${line.label ?? ''} ${line.text}'.trim()],
          ),
        );
        break;
      case RegulationLineKind.dash:
        blocks.add(
          RegulationBlock(
            type: RegulationBlockType.bullets,
            items: [line.text],
          ),
        );
        break;
      case RegulationLineKind.tableRow:
        blocks.add(
          RegulationBlock(
            type: RegulationBlockType.table,
            rows: [RegulationTableRow(cells: line.cells)],
          ),
        );
        break;
      case RegulationLineKind.labelValue:
        final label = line.label;
        final value = line.value;
        if (label != null && value != null) {
          blocks.add(
            RegulationBlock(
              type: RegulationBlockType.keyValues,
              entries: [RegulationEntry(label: label, value: value)],
            ),
          );
        }
        break;
      case RegulationLineKind.paragraph:
        blocks.add(
          RegulationBlock(
            type: RegulationBlockType.paragraph,
            text: line.text,
          ),
        );
        break;
      case RegulationLineKind.note:
        break;
    }
  }
  return blocks;
}

class RegulationSection {
  const RegulationSection({
    required this.code,
    required this.availability,
    required this.blocks,
  });

  final RegulationSectionCode code;
  final RegulationAvailability availability;
  final List<RegulationBlock> blocks;

  static RegulationSection? tryFromJson(Object? raw) {
    final json = _map(raw);
    if (json == null) return null;
    final code = RegulationSectionCode.tryParse(json['code']);
    final availability = RegulationAvailability.tryParse(json['availability']);
    if (code == null || availability == null) return null;
    final rawBlocks = json['blocks'];
    final blocks = rawBlocks is List
        ? rawBlocks
            .map(RegulationBlock.tryFromJson)
            .whereType<RegulationBlock>()
            .toList(growable: false)
        : const <RegulationBlock>[];
    if (availability == RegulationAvailability.present && blocks.isEmpty) {
      return null;
    }
    return RegulationSection(
      code: code,
      availability: availability,
      blocks: blocks,
    );
  }
}

class RegulationBlock {
  const RegulationBlock({
    required this.type,
    this.text,
    this.items = const [],
    this.entries = const [],
    this.columns = const [],
    this.rows = const [],
    this.divisions = const [],
  });

  final RegulationBlockType type;
  final String? text;
  final List<String> items;
  final List<RegulationEntry> entries;
  final List<String> columns;
  final List<RegulationTableRow> rows;
  final List<RegulationDivisionItem> divisions;

  static RegulationBlock? tryFromJson(Object? raw) {
    final json = _map(raw);
    if (json == null) return null;
    final type = RegulationBlockType.tryParse(json['type']);
    if (type == null) return null;
    final text = _text(json['text']);
    final items = _strings(json['items']);
    final entries = _objects(json['entries'], RegulationEntry.tryFromJson);
    final rows = _objects(json['rows'], RegulationTableRow.tryFromJson);
    final divisions = _objects(
      json['divisions'],
      RegulationDivisionItem.tryFromJson,
    );
    final columns = _strings(json['columns']);
    final hasContent = switch (type) {
      RegulationBlockType.paragraph ||
      RegulationBlockType.subheading ||
      RegulationBlockType.notice =>
        text != null,
      RegulationBlockType.bullets => items.isNotEmpty,
      RegulationBlockType.keyValues => entries.isNotEmpty,
      RegulationBlockType.table => rows.isNotEmpty,
      RegulationBlockType.divisionSchedule => divisions.isNotEmpty,
    };
    if (!hasContent) return null;
    return RegulationBlock(
      type: type,
      text: text,
      items: items,
      entries: entries,
      columns: columns,
      rows: rows,
      divisions: divisions,
    );
  }
}

class RegulationEntry {
  const RegulationEntry({required this.label, required this.value});

  final String label;
  final String value;

  static RegulationEntry? tryFromJson(Object? raw) {
    final json = _map(raw);
    if (json == null) return null;
    final label = _text(json['label']);
    final value = _text(json['value']);
    return label == null || value == null
        ? null
        : RegulationEntry(label: label, value: value);
  }
}

class RegulationTableRow {
  const RegulationTableRow({required this.cells});

  final List<String> cells;

  static RegulationTableRow? tryFromJson(Object? raw) {
    final json = _map(raw);
    if (json == null) return null;
    final cells = _strings(json['cells']);
    return cells.isEmpty ? null : RegulationTableRow(cells: cells);
  }
}

class RegulationDivisionItem {
  const RegulationDivisionItem({
    required this.name,
    this.date,
    this.venue,
    this.fee,
    this.account,
    this.capacity,
  });

  final String name;
  final String? date;
  final String? venue;
  final String? fee;
  final String? account;
  final String? capacity;

  static RegulationDivisionItem? tryFromJson(Object? raw) {
    final json = _map(raw);
    if (json == null) return null;
    final name = _text(json['name']);
    if (name == null) return null;
    return RegulationDivisionItem(
      name: name,
      date: _text(json['date']),
      venue: _text(json['venue']),
      fee: _text(json['fee']),
      account: _text(json['account']),
      capacity: _text(json['capacity']),
    );
  }
}

RegulationSectionCode sectionCodeForLegacyLabel(String label) {
  final value = label.toLowerCase().replaceAll(RegExp(r'[\s_\-·:/()]'), '');
  if (RegExp(r'참가부서|참가자격|출전규정|예외부서|시드기준|참가규모|경기종목').hasMatch(value)) {
    return RegulationSectionCode.eligibility;
  }
  if (RegExp(r'일시|일정|대회일|경기일|장소|경기장|개최지').hasMatch(value)) {
    return RegulationSectionCode.scheduleVenue;
  }
  if (RegExp(r'환불|취소|변경').hasMatch(value)) {
    return RegulationSectionCode.refundChanges;
  }
  if (RegExp(r'신청|접수|참가비|입금|계좌|결제|예금주').hasMatch(value)) {
    return RegulationSectionCode.registrationPayment;
  }
  if (RegExp(r'시상|상금|참가상품|기념품').hasMatch(value)) {
    return RegulationSectionCode.awards;
  }
  if (RegExp(r'경기방식|진행방식|운영|사용구|주최|주관|후원|협찬').hasMatch(value)) {
    return RegulationSectionCode.matchOperations;
  }
  if (RegExp(r'문의|연락|전화|담당|안내').hasMatch(value)) {
    return RegulationSectionCode.noticesContact;
  }
  return RegulationSectionCode.other;
}

Map<Object?, Object?>? _map(Object? value) => value is Map ? value : null;

String? _text(Object? value) {
  if (value is! String) return null;
  final text = value.trim();
  return text.isEmpty ? null : text;
}

List<String> _strings(Object? value) => value is List
    ? value.map(_text).whereType<String>().toList(growable: false)
    : const [];

List<T> _objects<T>(Object? value, T? Function(Object?) parse) => value is List
    ? value.map(parse).whereType<T>().toList(growable: false)
    : const [];
